plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Diagnostics.hs

```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE TupleSections         #-}
+module Ide.Plugin.Cabal.Diagnostics
+( errorDiagnostic
+, warningDiagnostic
+  -- * Re-exports
+, FileDiagnostic
+, Diagnostic(..)
+)
+where
+
+import qualified Data.Text              as T
+import           Development.IDE        (FileDiagnostic,
+                                         ShowDiagnostic (ShowDiag))
+import           Distribution.Fields    (showPError, showPWarning)
+import qualified Ide.Plugin.Cabal.Parse as Lib
+import           Language.LSP.Types     (Diagnostic (..),
+                                         DiagnosticSeverity (..),
+                                         DiagnosticSource, NormalizedFilePath,
+                                         Position (Position), Range (Range),
+                                         fromNormalizedFilePath)
+
+-- | Produce a diagnostic from a Cabal parser error
+errorDiagnostic :: NormalizedFilePath -> Lib.PError -> FileDiagnostic
+errorDiagnostic fp err@(Lib.PError pos _) =
+  mkDiag fp (T.pack "parsing") DsError (toBeginningOfNextLine pos) msg
+  where
+    msg = T.pack $ showPError (fromNormalizedFilePath fp) err
+
+-- | Produce a diagnostic from a Cabal parser warning
+warningDiagnostic :: NormalizedFilePath -> Lib.PWarning -> FileDiagnostic
+warningDiagnostic fp warning@(Lib.PWarning _ pos _) =
+  mkDiag fp (T.pack "parsing") DsWarning (toBeginningOfNextLine pos) msg
+  where
+    msg = T.pack $ showPWarning (fromNormalizedFilePath fp) warning
+
+-- | The Cabal parser does not output a _range_ for a warning/error,
+-- only a single source code 'Lib.Position'.
+-- We define the range to be _from_ this position
+-- _to_ the first column of the next line.
+toBeginningOfNextLine :: Lib.Position -> Range
```

Comment:
Should we have a version of this that works on LSP positions and put it with the other position utilities? Seems like a thing other people might want.

Reply:
Added a function to hls-plugin-api